File: woorireact/src/Chat/services/assistantConversationApi.js
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "";
const IMAGE_URLS_START = "[WOORI_IMAGE_URLS]";
const IMAGE_URLS_END = "[/WOORI_IMAGE_URLS]";
const HIDDEN_MESSAGE_MARKER = "[WOORI_HIDDEN_MESSAGE]";

async function request(path, options = {}) {
  const response = await fetch(`${API_BASE_URL}${path}`, options);
  if (!response.ok) {
    throw new Error(`Assistant conversation request failed: ${response.status}`);
  }

  if (response.status === 204) return null;
  return response.json();
}

export function createAssistantConversation(seniorId) {
  return request(`/api/assistant-conversations/senior/${seniorId}`, {
    method: "POST",
  });
}

export function fetchAssistantConversations(seniorId) {
  return request(`/api/assistant-conversations/senior/${seniorId}`);
}

export function fetchAssistantMessages(seniorId, conversationId) {
  const params = new URLSearchParams({ seniorId });
  return request(`/api/assistant-conversations/${conversationId}/messages?${params}`)
    .then((messages) => messages.map(decodeMessage));
}

export function saveAssistantMessage(seniorId, conversationId, message) {
  const params = new URLSearchParams({ seniorId });
  const content = encodeMessage(message);
  return request(`/api/assistant-conversations/${conversationId}/messages?${params}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      role: message.role,
      content,
    }),
  });
}

export function updateAssistantConversationTitle(seniorId, conversationId, title) {
  const params = new URLSearchParams({ seniorId });
  return request(`/api/assistant-conversations/${conversationId}?${params}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ title }),
  });
}

export function deleteAssistantConversation(seniorId, conversationId) {
  const params = new URLSearchParams({ seniorId });
  return request(`/api/assistant-conversations/${conversationId}?${params}`, {
    method: "DELETE",
  });
}

function encodeMessage(message) {
  const rawContent = String(message.content || "");
  const content = message.hidden
    ? `${HIDDEN_MESSAGE_MARKER}\n${rawContent}`
    : rawContent;
  const imageUrls = message.imageUrls || (message.imageUrl ? [message.imageUrl] : []);
  if (imageUrls.length === 0) return content;

  return `${content}\n\n${IMAGE_URLS_START}${JSON.stringify(imageUrls)}${IMAGE_URLS_END}`;
}

function decodeMessage(message) {
  const content = String(message.content || "");
  const markerPattern = new RegExp(
    `\\n*${escapeRegExp(IMAGE_URLS_START)}([\\s\\S]*?)${escapeRegExp(IMAGE_URLS_END)}\\s*$`
  );
  const match = content.match(markerPattern);
  let decoded = message;

  if (match) {
    try {
      const imageUrls = JSON.parse(match[1]);
      decoded = {
        ...message,
        content: content.replace(markerPattern, "").trim(),
        imageUrls: Array.isArray(imageUrls) ? imageUrls : [],
      };
    } catch {
      decoded = message;
    }
  }

  const decodedContent = String(decoded.content || "");
  if (!decodedContent.startsWith(HIDDEN_MESSAGE_MARKER)) return decoded;

  return {
    ...decoded,
    content: decodedContent.slice(HIDDEN_MESSAGE_MARKER.length).trimStart(),
    hidden: true,
  };
}

function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: woori_link/woori_link_flutter/lib/api/assistant_conversation_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../constants.dart';
import 'http_client.dart';

const _imageUrlsStart = '[WOORI_IMAGE_URLS]';
const _imageUrlsEnd = '[/WOORI_IMAGE_URLS]';
const _hiddenMessageMarker = '[WOORI_HIDDEN_MESSAGE]';

class AssistantConversationApi {
  static Future<Map<String, dynamic>> createConversation(int seniorId) async {
    final res = await http.post(
      Uri.parse('$baseUrl/assistant-conversations/senior/$seniorId'),
      headers: await authHeaders(),
    );
    if (res.statusCode == 200 || res.statusCode == 201) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw Exception('대화 생성 실패');
  }

  static Future<List<dynamic>> fetchConversations(int seniorId) async {
    final res = await http.get(
      Uri.parse('$baseUrl/assistant-conversations/senior/$seniorId'),
      headers: await authHeaders(),
    );
    if (res.statusCode == 200) return jsonDecode(res.body) as List<dynamic>;
    throw Exception('대화 목록 조회 실패');
  }

  static Future<List<Map<String, dynamic>>> fetchMessages(
    int seniorId,
    dynamic conversationId,
  ) async {
    final uri = Uri.parse('$baseUrl/assistant-conversations/$conversationId/messages')
        .replace(queryParameters: {'seniorId': '$seniorId'});
    final res = await http.get(uri, headers: await authHeaders());
    if (res.statusCode == 200) {
      final messages = jsonDecode(res.body) as List<dynamic>;
      return messages
          .map((message) => _decodeMessage(message as Map<String, dynamic>))
          .toList();
    }
    throw Exception('대화 메시지 조회 실패');
  }

  static Future<void> saveMessage(
    int seniorId,
    dynamic conversationId,
    Map<String, dynamic> message,
  ) async {
    final uri = Uri.parse('$baseUrl/assistant-conversations/$conversationId/messages')
        .replace(queryParameters: {'seniorId': '$seniorId'});
    final res = await http.post(
      uri,
      headers: await authHeaders(),
      body: jsonEncode({
        'role': message['role'],
        'content': _encodeMessage(message),
      }),
    );
    if (res.statusCode == 200 || res.statusCode == 201) return;
    throw Exception('대화 메시지 저장 실패');
  }

  static Future<void> updateTitle(
    int seniorId,
    dynamic conversationId,
    String title,
  ) async {
    final uri = Uri.parse('$baseUrl/assistant-conversations/$conversationId')
        .replace(queryParameters: {'seniorId': '$seniorId'});
    final res = await http.patch(
      uri,
      headers: await authHeaders(),
      body: jsonEncode({'title': title}),
    );
    if (res.statusCode == 200 || res.statusCode == 204) return;
    throw Exception('대화 제목 수정 실패');
  }

  static Future<void> deleteConversation(int seniorId, dynamic conversationId) async {
    final uri = Uri.parse('$baseUrl/assistant-conversations/$conversationId')
        .replace(queryParameters: {'seniorId': '$seniorId'});
    final res = await http.delete(uri, headers: await authHeaders());
    if (res.statusCode == 200 || res.statusCode == 204) return;
    throw Exception('대화 삭제 실패');
  }
}

String _encodeMessage(Map<String, dynamic> message) {
  final rawContent = '${message['content'] ?? ''}';
  final content = message['hidden'] == true
      ? '$_hiddenMessageMarker\n$rawContent'
      : rawContent;
  final imageUrls = message['imageUrls'] is List ? message['imageUrls'] as List : [];
  if (imageUrls.isEmpty) return content;
  return '$content\n\n$_imageUrlsStart${jsonEncode(imageUrls)}$_imageUrlsEnd';
}

Map<String, dynamic> _decodeMessage(Map<String, dynamic> message) {
  var decoded = Map<String, dynamic>.from(message);
  var content = '${decoded['content'] ?? ''}';

  final markerPattern = RegExp(
    r'\n*' +
        RegExp.escape(_imageUrlsStart) +
        r'([\s\S]*?)' +
        RegExp.escape(_imageUrlsEnd) +
        r'\s*$',
  );
  final match = markerPattern.firstMatch(content);
  if (match != null) {
    try {
      decoded['imageUrls'] = jsonDecode(match.group(1) ?? '[]');
      content = content.replaceFirst(markerPattern, '').trim();
    } catch (_) {
      // Keep raw content if legacy image metadata is malformed.
    }
  }

  if (content.startsWith(_hiddenMessageMarker)) {
    decoded['hidden'] = true;
    content = content.substring(_hiddenMessageMarker.length).trimLeft();
  }

  decoded['content'] = content;
  decoded['role'] = '${decoded['role'] ?? 'assistant'}'.toLowerCase();
  return decoded;
}
